zakirov.marat/S1

ZakirovMS opened the pull request:
Ring, Forward, Fake
7 баллов. Расширенный с итераторами, с перемещением и копированием, const-iterators
+iterators

Теперь по-модному выделяю голубым то, что сделано

![image](https://github.com/user-attachments/assets/0743f1de-6ea9-4a12-a7d7-78e1efcd9fb5)

ZakirovMS committed:
Initial commit for S1

ZakirovMS committed:
Initial for input/output data files

ZakirovMS committed:
Logic of the main function without implementing functions

ZakirovMS committed:
Implementation of obtaining a sequence

ZakirovMS committed:
Implementation using iterator list

ZakirovMS committed:
Implementation using copy of list

ZakirovMS committed:
The version of the program that is being compiled

ZakirovMS committed:
Almost the correct output

ZakirovMS committed:
Implementation output  with iterator list

ZakirovMS committed:
Continuing to write the output logic, adding the following pointers

ZakirovMS committed:
Change in checks of array

ZakirovMS committed:
CG fix

ZakirovMS committed:
Trailing whitespaces fix

ZakirovMS committed:
Fix of input, fix of output as a consequence

ZakirovMS committed:
Correcting trailing spaces

ZakirovMS committed:
Unused variable removed

ZakirovMS committed:
Fix of getting too large numbers and summ output implementation

ZakirovMS committed:
Fix of input only line names

ZakirovMS committed:
Fix of input shorter lines before longer one

ZakirovMS committed:
Adding a check for empty input

ZakirovMS committed:
Correct empty input according to criteria

ZakirovMS committed:
Unit test for main added

ZakirovMS committed:
Initial my list commit

ZakirovMS committed:
CG

ZakirovMS committed:
I removed start of implementation class

ZakirovMS committed:
renaming files

ZakirovMS committed:
C in russian keyboard and C in english keyboard is similar

Volgarenok commented:
> Ring, Forward, Fake

Не приложили выбранные критерии

ZakirovMS committed:
Starting implementation of fwd_list with fake and head field

ZakirovMS committed:
Creating Iterators for a forward list

ZakirovMS committed:
First column implementation

ZakirovMS committed:
Getting rid of the cyclical dependency between fwd_list.hpp and fwd_iterator.hpp, and other fixes

ZakirovMS committed:
Constructors of second and third column and insert_after implementation

ZakirovMS committed:
name variable typo fix

ZakirovMS committed:
Implementation of part of second column

ZakirovMS committed:
Full second column implemented

ZakirovMS committed:
Full third column implemented

ZakirovMS committed:
Fixes of typos in name of variables after replacement

ZakirovMS committed:
Replacing the standard list with a custom list. With a memory leak

ZakirovMS committed:
Trailing whitespaces fix

ZakirovMS committed:
Memory leak fix

ZakirovMS committed:
Trailing whitespaces fix

ZakirovMS committed:
Start of making a tests

ZakirovMS committed:
Test for constructors

ZakirovMS committed:
Fix of array_ptr

ZakirovMS committed:
Implementation of assigment operator tests

ZakirovMS commented:
Сейчас продолжаю прописывать тесты

ZakirovMS committed:
Merge branch 'master' into zakirov.marat/S1

Volgarenok commented on zakirov.marat/S1/io_containers.hpp:
```diff
@@ -0,0 +1,26 @@
+using list_pair = zakirov::FwdList< pair_strlist >;
+using list_pair_it = zakirov::FwdIterator< pair_strlist >;
+using list_iter = zakirov::FwdList< list_ull_it >;
+using list_iter_it = zakirov::FwdIterator< zakirov::FwdIterator< list_ull > >;
```
1. Глобальные псевдонимы в заголовочных файлах - это плохо
2. Псевдонимы итераторов вы должны получать из имени типа контейнера. Так это ерунда совсем

ZakirovMS replied:
1. Перенес псевдонимы в область видимости "zakirov"
2. Заменмл псевдонимы итераторов на auto для лучшей читаемости, псевдонимы удалил

Volgarenok commented on zakirov.marat/S1/io_containers.hpp:
```diff
@@ -0,0 +1,26 @@
+
+using list_ull = zakirov::FwdList< unsigned long long >;
+using list_ull_it = zakirov::FwdIterator< unsigned long long >;
+using it_to_it = zakirov::FwdIterator<list_ull_it>;
```
CG: обратите внимание на расстановку пробелов в шаблонах

ZakirovMS replied:
Шаблоны с ошибками были удалены в предыдщуем пункте

Volgarenok commented on zakirov.marat/S1/io_containers.cpp:
```diff
@@ -0,0 +1,155 @@
+
+    for (; iter_list != list_iterators.end(); ++iter_list)
+    {
+      if ((* iter_list) != (* pair_iter).second.end())
```
CG: обратите внимание на требования к расстановке пробелов у операторов

ZakirovMS replied:
Пробелы при разыменовывании итератора убраны, где это возможно, (* iter).method заменено на iter->method

Volgarenok commented on zakirov.marat/S1/io_containers.cpp:
```diff
@@ -0,0 +1,155 @@
+
+  list_iter list_iterators;
+  it_to_it iter_list = list_iterators.before_begin();
+  list_pair_it pair_iter = forward_list.begin();
```
Вот эта пачка псевдонимов ничего не даёт. Только скрывает понимание того, что происходит. Текст сливается. Когда мы говорим про итератор, нам важно понимать от какого контейнера - ожидается обращение к имени типа контейнера, а у вас псевдоним, который всё скрывает

ZakirovMS replied:
Лишние псевдонимы удалены

Volgarenok commented on zakirov.marat/S1/fwd_list_node.hpp:
```diff
@@ -0,0 +1,36 @@
+  FwdListNode< T >::FwdListNode():
+    data_(new T),
+    next_(nullptr)
+  {}
```
А зачем вам такая нода?

ZakirovMS replied:
Такой конструктор для Node создавался на случай, если пользователь захочет иметь узел без данных для его заполнения. Впрочем, никто Node напрямую не конструирует, так как это деталь реализации, а все действия с Node происходят через класс FwdList... Так что он бесполезен. Я удалил этот конструктор.

Volgarenok commented on zakirov.marat/S1/fwd_list_node.hpp:
```diff
@@ -0,0 +1,36 @@
+
+    FwdListNode();
+    FwdListNode(T data);
+    FwdListNode(T data, FwdListNode< T > * next);
```
Почему по значению?

ZakirovMS replied:
Добавил передачу по ссылке/константной ссылке, где это корректно

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+      }
+    }
+  }
+}
```
Явно же дублирование

ZakirovMS replied:
Избавился от копирования, переиспользовав remove_if в remove

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+    {
+      insert_after(inserter, *first);
+    }
+  }
```
Хочется строгую гарантию

ZakirovMS replied:
Обеспечил строгую гарантию с помощью реализованного ранее метода swap. Переиспользование assign(InputIterator, InputIterator) оставил.

Volgarenok replied:
А у оператора копирующего присваивания?

ZakirovMS replied:
Переиспользовал метод assigment для обеспечения строгой гарантии безопасности

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+    {
+      push_front(val);
+    }
+  }
```
Аналогично. Здесь тоже строгую

Volgarenok replied:
Если бы вы взяли аналогичный конструктор, то смогли бы переиспользовать другой assign

ZakirovMS replied:
После изменения реализации assign через конструирование и swap, мне показалось нецелесообразным переиспользовать в этой функции assign(InputIterator, InputIterator), так как в таком случае будет применяться еще один конструктор во втором assign, а без конструирования, как в случае assign(std::initializer_list< T >) реализовать метод не получится. Я использовал ту же самую форму "конструирование-swap", как в assign с итераторами.

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+  template< typename T >
+  void FwdList< T >::splice_after(FwdIterator< T > pos, FwdList && fwdlst, FwdIterator< T > first, FwdIterator< T > last)
+  {
+    FwdList & l_value_fwdlst = fwdlst;
```
Так он и так тут l-value. Зачем?

ZakirovMS replied:
Удалил

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+    FwdListNode< T > * temporary_node = other.fake_node_;
+    other.fake_node_ = fake_node_;
+    fake_node_ = temporary_node;
+  }
```
Можно стандартный swap использовать в реализации своего

ZakirovMS replied:
Использовал std::swap

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+
+    return counter;
+  }
+
```
Есть требование, чтобы размер вычислялся за О(1). Вариант с таким долгим вычислением не подойдёт нам

ZakirovMS replied:
Добавил поле size_ и логику изменения size_ в методы класса

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+    }
+
+    return 0;
+  }
```
Причём тут 1 и 0, когда у нас bool?

ZakirovMS replied:
Заменил на true-false

Volgarenok replied:
А зачем true/false, когда у вас есть результат логического выражения?

ZakirovMS replied:
Теперь возвращаю результат логического выражения

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+  template< typename T >
+  void FwdList< T >::push_front(const T && data)
+  {
+    FwdListNode< T > * new_element = new FwdListNode< T >(std::move(data));
```
Нет смысла в этой перегрузке. Если по ссылке на константу передано, то вы ничего и не переместите

ZakirovMS replied:
Убрал квалификатор const для реализации перемещения

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+  template< typename T >
+  FwdList< T > & FwdList< T >::operator=(FwdList< T > && fwdlst)
+  {
+    if (this != &fwdlst)
```
Для взятия адреса используйте `std::addressof`

ZakirovMS replied:
Использовал std::addressof() для получения адреса

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+  FwdList< T > & FwdList< T >::operator=(const FwdList< T > & fwdlst)
+  {
+    clear();
+    FwdIterator<T> inserter = begin();
```
CG: обратите внимание на расстановку пробелов в шаблонах

ZakirovMS replied:
Выставил пробелы по CG

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+    {
+      insert_after(inserter, *i);
+    }
+  }
```
Ну вот же у вас рядом конструктор с итераторами. Может его использовать?

ZakirovMS replied:
Переиспользовал конструктор с input_iterator в данном конструкторе и конструкторе FwdList(const FwdList &)

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+    void erase_after(FwdIterator< T > element);
+    void insert_after(FwdIterator< T > element, const T & data);
+    bool empty();
+    size_t size();
```
Может тут хотя бы noexcept поставить?

ZakirovMS replied:
Добавил квалификаторы const и noexcept в методы

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+namespace zakirov
+{
+  template< typename T >
+  class FwdList
```
Что-то вы совсем нигде const не поставили практически

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+    FwdList< T > & operator=(const FwdList< T > & fwdlst);
+    FwdList< T > & operator=(FwdList< T > && fwdlst);
+    FwdList< T > & operator=(std::initializer_list< T > init_list);
+    FwdListNode< T > * front();
```
Узлы - это детали реализации. Как вы их возвращаете-то?

ZakirovMS replied:
Все методы, в которых я придумал возвращать узлы, теперь возвращают итераторы

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+    void assign(std::initializer_list< T > init_list);
+    void clear();
+    void remove(T data);
+    template< typename P>
```
CG

ZakirovMS replied:
Добавил пробел в шаблоне

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+    void assign(InputIterator first, InputIterator last);
+    void assign(std::initializer_list< T > init_list);
+    void clear();
+    void remove(T data);
```
Почему по значению?

ZakirovMS replied:
Изменил на передачу по константной ссылке

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+    template< typename InputIterator >
+    void assign(InputIterator first, InputIterator last);
+    void assign(std::initializer_list< T > init_list);
+    void clear();
```
Ну в clear-то наверное точно никак исключений быть не может (во всяком случае не должно быть)

ZakirovMS replied:
Добавил noexcept

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,395 @@
+#ifndef MY_LIST_HPP
+#define MY_LIST_HPP
```
"My"? Только вам им можно пользоваться?

ZakirovMS replied:
Я сам понял, что название не очень, но пропустил его в header guard. Исправил.

Volgarenok commented on zakirov.marat/S1/fwd_iterator.hpp:
```diff
@@ -0,0 +1,142 @@
+  struct CFwdIterator: public std::iterator< std::forward_iterator_tag, T >
+  {
+    CFwdIterator();
+    CFwdIterator(FwdListNode< T > * node);
```
Нужно ограничить участие этого конструктора в неявных преобразованиях

ZakirovMS replied:
Добавил квалификатор explicit для исключения неявных преобразований в работе конструктора

Volgarenok commented on zakirov.marat/S1/fwd_iterator.hpp:
```diff
@@ -0,0 +1,142 @@
+  struct CFwdIterator: public std::iterator< std::forward_iterator_tag, T >
+  {
+    CFwdIterator();
+    CFwdIterator(FwdListNode< T > * node);
```
Этот конструктор является деталью реализации. Его не должно быть в открытом интерфейсе

ZakirovMS replied:
Детали реализации инкапсулированы, FwdList теперь friend class для FwdIterator

Volgarenok commented on zakirov.marat/S1/fwd_iterator.hpp:
```diff
@@ -0,0 +1,142 @@
+    FwdIterator(const FwdIterator< T > &) = default;
+    FwdIterator< T > & operator=(const FwdIterator< T > &) = default;
+    T & operator*();
+    T * operator->();
```
Константные версии всё равно должны быть

ZakirovMS replied:
Добавил константные версии

Volgarenok commented on zakirov.marat/S1/fwd_iterator.hpp:
```diff
@@ -0,0 +1,142 @@
+    FwdIterator< T > & operator++(int);
+    bool operator!=(const FwdIterator< T > & other) const noexcept;
+    bool operator==(const FwdIterator< T > & other) const noexcept;
+    zakirov::FwdListNode< T > * node_;
```
Поля наверное не должны быть открытыми. Как считаете?

ZakirovMS replied:
Перенес поля в private область видимости

ZakirovMS committed:
Merge branch 'master' into zakirov.marat/S1

ZakirovMS committed:
Constancy addition to list

ZakirovMS committed:
1. Moving aliases to Zakirov scope
2. Removing iterator aliases

ZakirovMS committed:
Merge branch 'master' into zakirov.marat/S1

ZakirovMS committed:
1. Move the constructor, which is an implementation detail, to private
2. Add noexcept for iterator methods

ZakirovMS committed:
1. Implementation details are encapsulated
2. FwdList is a friend class for FwdIterator

ZakirovMS committed:
1. Removed incorrect node constructor
2. Added pass by reference where correct
3. Fixed duplication in remove_if and remove

ZakirovMS committed:
1. std::swap is used in swap method
2. swap method is used in assign for strong guarantee
3. Removed unnecessary variable creations

ZakirovMS committed:
1. Adding the size field
2. Correcting reuses, typos and CG

ZakirovMS committed:
Explicit in iterator

ZakirovMS committed:
Fix for const in splice_after, needed due to adding size_ field

ZakirovMS committed:
1. R-value assigned test is correct now

ZakirovMS committed:
Constancy method -> and * was added

ZakirovMS committed:
Complete testing of FwdList class completed

ZakirovMS committed:
Merge branch 'master' into zakirov.marat/S1

ZakirovMS committed:
1. Addition noexcept to splice_after

ZakirovMS committed:
Noexcept for assign was a typo

Volgarenok commented on zakirov.marat/S1/fwd_list.hpp:
```diff
@@ -0,0 +1,383 @@
+    {
+      swap(fwdlst);
+    }
+
```
Получается перемещение - это проверка и обмен?.. Ну нет же

ZakirovMS replied:
Теперь в результате работы конструктора привожу перемещаемый объект к безопасному (базовому) виду

ZakirovMS committed:
1. Reused assign method to provide strong safety guarantee in copy assignment operator, added check
2. Write corresponding test for check
3. Returns result of logical expression in empty method
4. As a result of work of move constructor, moved object is transferred to safe state

ZakirovMS committed:
Merge branch 'master' into zakirov.marat/S1
